### PR TITLE
New Luhn instant answer

### PR DESCRIPTION
**Pull request by regagain**
https://duck.co/ia/view/luhn_algorithm

###### What does your Pull Request do (check all that apply)?

Choose the most relevant items and use the following title template to name
your Pull Request.

- [x] New Instant Answer
    - [ ] Cheat Sheets: **`New {Cheat Sheet Name} Cheat Sheet`**
    - [x] Other: **`New Luhn Instant Answer`**

###### Description of changes

New IA, calculating Luhn checksum.

---

Instant Answer Page: https://duck.co/ia/view/luhn_algorithm

### CODE REVIEW

**Comment by daxtheduck** [comment]
@regagain Hey!
        
Thanks for taking the time to contribute! We really appreciate it.

We work closely with every contributor to make Instant Answers the best they can be, so we appreciate your patience as we look over your code. From here, the process usually goes like this:

1. Pull Request is reviewed by the DuckDuckGo staff and community
2. Staff and community will leave feedback with any necessary updates to the function or design.
3. Once you've made any necessary corrections, then your Instant Answer will be merged and deployed live on DuckDuckGo!

If you have any questions along the way, feel free to ask them here. Our staff and community are also available on Slack to answer any questions you may have. If you'd like to join us there please head to https://quackslack.herokuapp.com/ to get an invite.

>More Info: http://docs.duckduckhack.com/submitting/submitting-overview.html

Thanks!

**Comment by daxtheduck** [comment]
Metadata placeholder post
            
*If you haven't already, please edit your first comment to include a link to the IA page at the bottom.*
            
---
            
**How to find your IA page**
- New Instant Answer
  -- follow the steps [here](https://duck.co/ia/new_ia) to create your IA page.
              
- Existing Instant Answer
  -- find your IA page from this [list](https://duck.co/ia)

File: lib/DDG/Goodie/Luhn.pm [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+handle remainder => sub {
+
+    return unless /^\s*\d+(?:(?:\s|,)+\d+)*\s*$/;
+    
+    my $tmp = $_ =~ s/\s//gr; # removing all white spaces
+
+    my $result = check_digit($tmp);
+
+    return "The Luhn check digit of $_ is $result.",
+      structured_answer => {
```
Comment (mintsoft) [inline lib/DDG/Goodie/Luhn.pm]: We're now moving towards full templates for all Goodies; can you switch this to the Text Template? There's an example here: https://github.com/duckduckgo/zeroclickinfo-goodies/blob/master/lib/DDG/Goodie/Stardate.pm#L22

File: lib/DDG/Goodie/Luhn.pm [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+# Caching - http://docs.duckduckhack.com/backend-reference/api-reference.html#caching`
+zci is_cached => 1;
+
+# Triggers - http://docs.duckduckhack.com/walkthroughs/calculation.html#triggers
+triggers startend => 'luhn';
+
+# Handle statement
+handle remainder => sub {
+
+    return unless /^\s*\d+(?:(?:\s|,)+\d+)*\s*$/;
```
Comment (GuiltyDolphin) [inline lib/DDG/Goodie/Luhn.pm]: Lets use the NumberStyle GoodieRole for parsing and displaying the numbers.

File: lib/DDG/Goodie/Luhn.pm [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+    
+    my $tmp = $_ =~ s/\s//gr; # removing all white spaces
```
Comment (GuiltyDolphin) [inline lib/DDG/Goodie/Luhn.pm]: We shouldn't need to do this - `remainder` will filter white space.

Comment (GuiltyDolphin) [inline lib/DDG/Goodie/Luhn.pm]: We shouldn't need to do this - `remainder` will filter white space.

File: lib/DDG/Goodie/Luhn.pm (hunk 2 above)
Comment (GuiltyDolphin) [inline lib/DDG/Goodie/Luhn.pm]: Also, how about using a guard so we can trigger on `luhn check digit of ...`.

Comment (GuiltyDolphin) [inline lib/DDG/Goodie/Luhn.pm]: Also, how about using a guard so we can trigger on `luhn check digit of ...`.

Comment (regagain) [inline lib/DDG/Goodie/Luhn.pm]: Thanks for the feedback, what is the best way to do this? Is to add a start trigger or something else?

File: lib/DDG/Goodie/Luhn.pm (hunk 3 above)
Comment (regagain) [inline lib/DDG/Goodie/Luhn.pm]: @GuiltyDolphin It seems it doesn't remove white spaces inside the query, does it? Queries like "luhn 123 123" are failing when I remove this.

**Comment by zekiel** [comment]
@regagain this is going up on beta.duckduckgo.com for testing. Can you please add more examples queries and description to [your Instant Answer page](https://duck.co/ia/view/luhn_algorithm) so others can test it more thoroughly?

File: t/Luhn.t [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+            }
+        };
+}
+
+sub build_test { test_zci(build_structured_answer(@_)) }
+
+ddg_goodie_test(
+    [qw( DDG::Goodie::Luhn )],
+    'luhn 1' => build_test('8', 'The Luhn check digit of 1 is 8.'),,
+    'luhn 4242 4242 424' => build_test('2', 'The Luhn check digit of 4242 4242 424 is 2.'),
```
Comment (gaulrobe) [inline t/Luhn.t]: Because of the whitespace trimming being done on https://github.com/regagain/zeroclickinfo-goodies/blob/11f822a1e5721bade48819b08ccfe7e66881bfb2/lib/DDG/Goodie/Luhn.pm#L23 this is really calculating the Luhn check-digit of 42424242424. Do we really want to trigger this in cases like this, or should we restrict the trigger to `luhn [some integer]` only?

Comment (regagain) [inline t/Luhn.t]: Since:
1. Luhn is used for, amongst other things, credit cards numbers, and that these numbers are often written with spaces.
2. Spaces do not change the result in any way.
I think it makes sense to trigger in cases like this.

Comment (gaulrobe) [inline t/Luhn.t]: I can _buy_ that :stuck_out_tongue_winking_eye: 

Thanks for the clarification!

File: lib/DDG/Goodie/Luhn.pm [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+# ABSTRACT: Calculate check digit according to Luhn formula. 
+
+use DDG::Goodie;
+use strict;
+use Algorithm::LUHN qw/check_digit /;
+with 'DDG::GoodieRole::NumberStyler';
+
+zci answer_type => 'luhn';
+
+# Caching - http://docs.duckduckhack.com/backend-reference/api-reference.html#caching`
```
Comment (mintsoft) [inline lib/DDG/Goodie/Luhn.pm]: Can you remove these comments? They're largely for instruction

File: lib/DDG/Goodie/Luhn.pm [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+zci is_cached => 1;
+
+# Triggers - http://docs.duckduckhack.com/walkthroughs/calculation.html#triggers
+triggers startend => 'luhn';
+
+# Handle statement
+handle remainder => sub {
+
+    my $number_re = number_style_regex();
```
Comment (mintsoft) [inline lib/DDG/Goodie/Luhn.pm]: Can you define that outside `handle` so it's only built up once :)

File: lib/DDG/Goodie/Luhn.pm [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+    return unless /^$number_re$/;
+    
+    my $tmp = $_ =~ s/\s//gr; # removing all white spaces
+
+    my $result = check_digit($tmp);
+
+    return $result,
+      structured_answer => {
+        data => {
+          title => $result,
```
Comment (mintsoft) [inline lib/DDG/Goodie/Luhn.pm]: Nit: Can we use 4 spaces instead of 2?

File: t/Luhn.t (hunk 4 above)
Comment (mintsoft) [inline t/Luhn.t]: Can we add a test for the same input values without the whitespace so it's obvious from the test suite that the behaviour is that the whitespace can be ignored?

File: t/Luhn.t [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+use warnings;
+use Test::More;
+use Test::Deep;
+use DDG::Test::Goodie;
+
+zci answer_type => "luhn";
+zci is_cached   => 1;
+
+sub build_structured_answer {
+    my ($result, $formatted) = @_;
```
Comment (mintsoft) [inline t/Luhn.t]: Can you switch this out to be something like `my ($result, $input) = @_' `
and then use:
```perl
....
data => {
    title => $result,
    subtitle => "The Luhn check digit of $input is $result."
}
....

**Comment by mintsoft** [comment]
Thanks @regagain I've put some more feedback against the diff :)

**Comment by moollaza** [comment]
@regagain Ping! I'm just checking in to see if you're still interested in finishing this one up? We'd love to see this go live!

@regagain, @mintsoft, It looks like only some minor edits are left?

There's absolutely no rush, just wanted to make sure you're not stuck on anything.

**Comment by regagain** [comment]
@moollaza I appreciate the followup! :+1:  I was away for a while, but I will get back to this soon (most likely in the coming days).

**Comment by regagain** [comment]
@moollaza @mintsoft I think I am done! Let me know if I have forgotten anything.
And of course, feedback is welcome!

**Comment by tagawa** [comment]
Thanks for coming back to this @regagain. I've just deployed the latest version to beta for testing: https://beta.duckduckgo.com/?q=luhn+12345&ia=answer
Looking good to me.

**Comment by mintsoft** [comment]
@tagawa LGTM to me too; although I've just noticed that the parcelforce IA seems a little keen: https://beta.duckduckgo.com/?q=luhn+1234554651&ia=answer

**Comment by tagawa** [comment]
Well spotted @mintsoft. I've filed an issue for that: https://github.com/duckduckgo/zeroclickinfo-goodies/issues/3391

Incidentally this IA is now not triggering for me on beta. I'm trying to find out why not.

**Comment by tagawa** [comment]
It's back on beta. It might have been me that removed it by accident 😵 
Anyway, no further feedback so let's get merging.

Thank you again @regagain! I love seeing conversion/utility Instant Answers like this.

**Comment by regagain** [comment]
Thank you guys for the support and feedback!
